README.md
# Ubuntu Post Installation Script : after-effects

Effortless way to setup your fresh Ubuntu, Linux Mint, Debian installs. Please read the [FAQ](https://ae.prasadt.com/faq/dependencies/) & the [Docs](https://ae.prasadt.com/getting-started/)
before trying out.

[![Build Status](https://travis-ci.org/tprasadtp/ubuntu-post-install.svg?branch=master)](https://travis-ci.org/tprasadtp/ubuntu-post-install)
[![Latest Version](https://img.shields.io/badge/dynamic/json.svg?label=Version&style=flat&url=https://ae.prasadt.com/api/json/version.json&query=version.number&prefix=v)](https://github.com/tprasadtp/ubuntu-post-install/)

## URLs

```bash
Github URL: https://github.com/tprasadtp/ubuntu-post-install
Github URL: https://github.com/tprasadtp/after-effects-ubuntu
```

## How to use this?

Install your choice of Ubuntu/Debian or its derivative as you would( If you wish to automate that too, you can use `preseed.cfg file`)

### Step 1: Get the script

#### Without Git

Run this in Terminal

```sh
wget -Nnv https://raw.githubusercontent.com/tprasadtp/ubuntu-post-install/master/get-after-effects.sh -O - | bash
```

Or if you want to use YAML configuration & not download or use .list files

```sh
wget -Nnv https://raw.githubusercontent.com/tprasadtp/ubuntu-post-install/master/get-after-effects.sh -O - | bash -s -- -r
```

#### With Git

If you already have git on your system already you can use,

```bash
git clone --depth 1 https://github.com/tprasadtp/ubuntu-post-install.git && cd ubuntu-post-install
```

If you already have cloned the repo, you can use git to get the latest changes.

### Step 2: Update the lists or your YAML configuration file

Update the list files or YAML file to suit your needs. Change PPAs, add or delete packages to list, tweak flags etc.
Please see [Configuration](https://ae.prasadt.com/config/#package-lists) & [Tasks](https://ae.prasadt.com/tasks/#what-can-it-do).

### Step 3: Run the script

Run the script as **root**. You will get an error if you do not run the script as root.

- To use default YAML configuration, which can be found here `https://ae.prasadt.com/api/config.yml` Recommended
  ```sh
    sudo ./after-effects -Y
  ```

- To use local list files in `data` [WIP]
```sh
  sudo ./after-effects -L
```

> If you are running this in a docker container, you probably are root and its possible that you might be missing `sudo`. So In that case just run it as `./after-effects`. Be warned! You might be missing several dependencies of the script!!

### Step 4: Contribute!

Contribute tweaks, improvements ideas back upstream.
Please note that `gh-pages` branch is complete derivative of master. Please checks docs directory for documentation. This project uses `mkdocs` for documentation. You can use docker-compose to test documentation site.

## FAQ & Documentation

See /docs or visit [https://after-effects.prasadt.com](https://after-effects.prasadt.com).

## Security
Well, this isn't very secure or meant to be.
- You can verify them manually or via `./after-effects --verify`
- GPG signature of the script is also available at `api/gpg?version="VERSION_NUMBER"`

## Supported Distributions

A Complete  list of supported distributions is given below.

 Distribution    | Code name/Version |  Supported  | Notes
:----------------|:------------------|-------------|:------
Ubuntu 18.04     | Bionic Beaver     | Yes         |
Ubuntu 16.04     | Xenial Xerus      | Yes         |
Ubuntu 14.04     | Trusty Thar       | Yes         |
Ubuntu 18.10     | Cosmic Cuttlefish | --          | Only use it for testing
Ubuntu 17.10     | Artful Aardvark   | No          | Reached EOL July 2018
Ubuntu 17.10     | Zesty Zapus       | No          | Reached EOL Jan 13th 2018
Linux Mint 17    | Quina             | Yes         |
Linux-Mint 17.1  | Rebecca           | Yes         |
Linux-Mint 17.2  | Rafaela           | Yes         |
Linux-Mint 17.3  | Rosa              | Yes         |
Linux Mint 18    | Sarah             | Yes         |
Linux-Mint 18.1  | Serena            | Yes         |
Linux-Mint 18.2  | Sonya             | Yes         |
Linux-Mint 18.3  | Sylvia            | Yes         |
Linux Mint 19    | Tara              | Yes         | Not tested
Elementary-OS    | Freya             | Yes         | Not tested
Elementary-OS    | Loki              | Yes         | Not tested
Elementary-OS    | Juno              | Yes         | Not tested
Budgie Remix     | 16.04             | Yes         |
PoP! OS          | 17.10 & 18.04     | Yes         | Not tested
Linux-Lite       | 3.X               | Yes         | (Based on Ubuntu 16.04)
Zorin-OS         | 12.X              | Yes         | (Based on Ubuntu 16.04)
Bodhi Linux      | 4                 | Yes         | (Based on Ubuntu 16.04)
Debian 8         | Jessie            | Yes         |
Debian 9         | Stretch           | Yes         |

## Oh no! It crashed :(

- Well that shouldn't have happened. If it does, see [Script Crashes](https://ae.prasadt.com/faq/crash/) and consider opening an issue on Github.
- Please include the log file and terminal output if you can in the issue..
- **Never** set environment variable `CI=true` & `TRAVIS=true` unless you are running CI tests or are sure of its effects.

## Contributing & Forks

See [Contributing and forks](./.github/CONTRIBUTING)

## Credits

[![Analytics](https://ga-beacon.prasadt.com/UA-101760811-3/github/ubuntu-post-install?flat)](https://prasadt.com/google-analytics-beacon)


Changelog.md
# Changelogs

## _v4.0_

- Pre and Post Hooks for scripts. PS: Later thought: [WHY DID I DO IT?? WITH JUST BASH ITS PIA]
- Custom list of scripts can be run Before and after all the tasks.

## _v3.6_

- Add Option to install system wide python packages using pip
- Promote Bionic to stable and update codenames
- Update pre-release to cosmic, rename test script for pre release
- Remove unused dockerfiles

## _v3.4_

- Delete cached and downloaded DEBs. (Can be changed with `-k` flag )
- Remove Gnome Specific stuff from default list
- Remove PPAs incompatible with bionic
- Bionic is supported with --pre-release flag only with default list.

## _v3.3_

- Add Signal Desktop repository
- Better way to handle EOL error messages
- Fix: Visual Studio Code GPG key not deleted while removing repos.
- Lint Readme Markdown
- Nuke Submodules

## _v3.2_

- Allow Bionic test to fail on Travis.
- Only print logs in Travis if there is an error or a flag is passed.
- Switch to submodules for data directory
- List files have their own repo now.
- Zesty reaches EOL soon. Remove it.
- Use daily images for bionic

## _v3.1_

### Added Support for Ubuntu Bionic Beaver

- Added Support for Upcoming Ubuntu release bionic.
- Added an option to use repository for last stable release on bionic.
- Use Ubuntu Base 18.04 LTS (Bionic Beaver) daily build to build docker image.
- Allow Bionic tests to fail on Travis CI.
- Dockerfiles & tests for bionic.
- Inform in script if running on Upcoming release.
- Drop google-cloud-sdk from fix_repo_not_available. Use `--pre-release` if using beta/alpha Ubuntu release.
- Add Visual studio to repos instead of deb files
- Rename logging directory to after-effects

## _v3.0_

- ### Confirmation dialogs
  - Add Confirmation dialog using whiptail for Actions like Adding PPA, Repositories, Installing Apps, and Deb files
  - Provide an option via command line to bypass the confirmation dialog for ci and automated environments or when its too annoying

- ### Improvements to Simulate option
  - If the simulate option is selected then Only calculate the upgrade but do not perform upgrade.
  - Added improved Simulate options. Helpful when just want to change somethings and test scripts without really downloading and installing packages
  - Simulate option is by default false and can be toggled by passing `-s` or `--simulate` while running the script via command line
- ### Partner repositories and Derivatives
  - Do not Enable Canonical Partner repositories in Ubuntu derivatives as they are enabled in installer or are different than Ubuntu. This leaves Partner repositories as they were before
- ### Add Feature: Purge Unwanted Packages
  - Fix a bug in PPA purge function where it would wait for user confirmation
  - Add a feature to purge not required packages, usually games and boaltware which comes preloaded  with the install. It is necessary to pass a flag `-d` via command line to purge them while running the script since its dangerous
- ### Logs & Console messages
  - Improved logging . Redirecting errors and adding time-stamps works better.
- ### Travis CI and Docker Testing
  - Added Dockerfiles used for test cases since base Ubuntu image does not have few necessary packages (`ping, wget, lsb-release, whiptail, iputils-ping`) and configuration required. Since most of these are bound to present on Ubuntu/Ubuntu Server/ Ubuntu derivatives it is not necessary to add them in the script.
  - Added Docker and Travis CI detection . Now the logs will indicate if the script is running in Container or Travis CI
  - Test Script on Travis CI in parallel using jobs.
  - Test Script on Trusty using Travis Host
  - Test Script on Xenial, Artful and Zesty using Docker image build using Dockerfiles.
- ### Others
  - Fixed Bugs
  - Fixed Typos and spell errors
  - Fixed a bug where exiting script via pressing escape would cause message to be printed twice
  - Improved exit status handling

## _v2.0_

- Complete rewrite from scratch
- Improvements in logging and console output
- Reduced verbosity in terminal output
- Flexible with packages and deb files
- Reduced complex dependencies
- Easy to configure
- Add Simulate install option for installing deb files and apps. Easier to test scripts now .

## _v1.0_

- Initial upload.


get-after-effects.sh
#!/usr/bin/env bash
# shellcheck disable=SC2059
# Legal Preamble:
#
# This script is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; version 3.
#
# This script is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, see <https://www.gnu.org/licenses/gpl-3.0.txt>

# Author: Prasad Tengse
# Licence: GPLv3
# Github Repository: https://github.com/tprasadtp/ubuntu-post-install
# Requirements - Bash v4.4 and above
#              - Ubuntu 14.04/16.04/17.04/17.10 or Elementary Freya/Loki/Juno or Mint 17.X or 18.X
#              - Support for Ubuntu 18.04 Bionic is Experimental. Since it is still under development
#              - it is advised not to use it as daily driver and should be considered buggy.
#              - whiptail, lsb-release, wget

#
set -o pipefail
readonly SCRIPT=$(basename "$0")
readonly dir=$(cd -P -- "$(dirname -- "$0")" && pwd -P)
#colors for display
#readonly BLUE=$(tput setaf 6)
readonly YELLOW=$(tput setaf 3)
readonly NC=$(tput sgr 0)
ignore_git_folder="false"
GET_BASE_URL="https://raw.githubusercontent.com/tprasadtp/ubuntu-post-install/master"
GET_LIST_BASE_URL="https://raw.githubusercontent.com/tprasadtp/ubuntu-post-install/master/data"
GET_YAML_BASE_URL="https://new-badges--ubuntu-post-install.netlify.com"
function check_dependencies()
{
  #Function to check is dependencies are available
  local get_after_effects_dependencies=(wget)
  local dep_res
  printf "${YELLOW}Checking dependencies${NC}\n"
  for dep in "${get_after_effects_dependencies[@]}"; do
   command -v "$dep" > /dev/null
   dep_res=$?
   if [ "$dep_res" -eq 1 ]; then
     printf "${YELLOW}[ ERROR ] $dep is not installed!${NC}\n"
     exit -1;
   fi
  done
}

function get-after-effects()
{
  # Function to get after-effects main module
  if [ -d .git ] && [ "$ignore_git_folder" != "true" ]; then
    printf "This directory seems to be a git repository. Please use git pull or git fetch to update the script.\n"
    printf "If its not, please delete the .git folder and try again. Use -f opion to ignore this.\n"
  elif [ -d .git ] && [ "$ignore_git_folder" == "true" ]; then
    printf "This directory seems to be a git repository.\nSince --force is used, Cleaning up...\n"
    rm -f ./* ./.* ./docs/**/*.* ./docs/*.* ./api/* ./data/* ./tests/* ./vendor/* ./.vscode/* ./dockerfiles/* ./.github/* ./after-effects ./after-effects.* ./*.mlist ./*.mlist.*
  else
    printf "${YELLOW}Removing old files...${NC}\n"
    rm -f after-effects ./after-effects.* get.mlist .data/*.list ./get.mlist.* ./README.md ./README.md.* ./data/*.list ./data/*.list.*
  fi

  echo "Getting: after-effects"
  wget -q  "${GET_BASE_URL}"/after-effects
  printf "${YELLOW}Changing file permissions...${NC}\n"
  chmod +x ./after-effects
  if [ "$use_yaml" == "false" ]; then
    printf "${YELLOW}Getting Data and Lists...${NC}\n"
    wget -q "${GET_LIST_BASE_URL}"/get.mlist
    mkdir -p data
    while IFS= read -r line
      do
        echo "Getting: $line"
        wget -q -P ./data/ "${GET_LIST_BASE_URL}"/"$line"
      done <get.mlist
  else
    wget -q "${GET_YAML_BASE_URL}/api/config.yml" -O example-config.yml
    wget -q "${GET_YAML_BASE_URL}/api/version.yml" -O version.yml
    printf "[   Info  ] Be sure to pass the right flags while running the script\n"
  fi
  printf "${YELLOW}Please Run the script after-effects as root\n"
  printf "${YELLOW}sudo ./after-effects\n"
  printf "${YELLOW}For documentation visit: https://ae.prasadt.com${NC}\n"
}


function main()
{
  while [ "$1" != "" ]; do
        case ${1} in
            -l | --list )           use_yaml=false;
                                    printf "[   Info  ] Not Downloading list. Using YAML configurations.\n"
                                    ;;
            -f | --force )          ignore_git_folder=true
                                    printf "[   Info  ] Will ignore existing \'.git\' folder.\n"
                                    ;;
                * )                 printf "[  Error! ] Invalid option: $1\n"
                                    exit 1;
                                    ;;
      esac
    shift
  done

  check_dependencies
  get-after-effects
}

main "$@"
